### PR TITLE
detect service.version based on jar file name

### PR DESCRIPTION
**Pull request by zeitlinger**
Add `JarServiceVersionDetector`

~~Note: This PR already has a bug - see https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10540. Fixed now.~~

### CODE REVIEW

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java [hunk 1]
```diff
@@ -121,10 +149,27 @@ private Path pathIfExists(String programArguments) {
     return fileExists.test(candidate) ? candidate : null;
   }
 
-  private static String getServiceName(Path jarPath) {
+  private static NameAndVersion getServiceNameAndVersion(Path jarPath) {
     String jarName = jarPath.getFileName().toString();
     int dotIndex = jarName.lastIndexOf(".");
-    return dotIndex == -1 ? jarName : jarName.substring(0, dotIndex);
+    if (dotIndex == -1 || ANY_DIGIT.matcher(jarName.substring(dotIndex)).find()) {
```
[the remark this answers is not in the document]
Comment (jeanbisutti) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java:155]: A user could also potentially add a dot to the application name. For example: `my-service2.3-1.0.0.jar.` It's odd but feasible. So, I am not sure about the version extraction for the JAR name.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java [hunk 2]
```diff
@@ -121,10 +149,27 @@ private Path pathIfExists(String programArguments) {
     return fileExists.test(candidate) ? candidate : null;
   }
 
-  private static String getServiceName(Path jarPath) {
+  private static NameAndVersion getServiceNameAndVersion(Path jarPath) {
```
[the remark this answers is not in the document]
Comment (jeanbisutti) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java:152]: Perhaps after having solved https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10514#discussion_r1485967114? To be sure that this PR makes sense.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java:155]: it actually works :smile:  - see the tests

[the remark this answers is not in the document]
Comment (jeanbisutti) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java:155]: It's a bit strange for me to deduce a version from the Spring Boot JAR file name. [With the Spring Boot Maven/Gradle plugin, it's possible to override the default JAR file name.](https://docs.spring.io/spring-boot/docs/current/maven-plugin/reference/htmlsingle/#packaging.examples.custom-name) Let's imagine the user configures a JAR file name `application-2.jar`. How could we know  that the application name is `application-2`, or `application` with a `2` version number?

[the remark this answers is not in the document]
Comment (jeanbisutti) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java:155]: I would suggest to remove the Spring detector based on the JAR file name, and fallback as a last resort to the `build-info.properties` file or the manifest file. See https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10515#issuecomment-1938675839.

[the remark this answers is not in the document]
Comment (zeitlinger) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/JarServiceNameDetector.java:155]: are you saying that `unknown_service:java` would be better than a name based on the jar file?

**Comment by zeitlinger** [comment]
SIG meeting outcome: we'll wait until someone asks for this feature.

**Comment by zeitlinger** [comment]
meeting decision: not worth the complexity

**Pull request closed without merging**